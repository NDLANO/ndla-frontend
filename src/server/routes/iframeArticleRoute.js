/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import defined from 'defined';
import { Helmet } from 'react-helmet';
import { INTERNAL_SERVER_ERROR, OK } from 'http-status';

import { getHtmlLang, getLocaleObject } from '../../i18n';
import { fetchResourceTypesForResource } from '../../containers/Resources/resourceApi';
import IframePageContainer from '../../iframe/IframePageContainer';
import config from '../../config';
import handleError from '../../util/handleError';
import { renderPageWithData, renderHtml } from '../helpers/render';

const assets =
  process.env.NODE_ENV !== 'unittest'
    ? require(process.env.RAZZLE_ASSETS_MANIFEST) //eslint-disable-line
    : {
        client: { css: 'mock.css' },
        embed: { js: 'mock.js' },
        polyfill: { js: 'mock.js' },
        mathJaxConfig: { js: 'mock.js' },
      };

if (process.env.NODE_ENV === 'unittest') {
  Helmet.canUseDOM = false;
}

const getAssets = () => ({
  css: assets.client.css,
  js: [{ src: assets.embed.js }],
  polyfill: { src: assets.polyfill.js },
  mathJaxConfig: { js: assets.mathJaxConfig.js },
});

async function doRenderPage(initialProps) {
  const Page = config.disableSSR ? (
    ''
  ) : (
    <IframePageContainer {...initialProps} />
  );
  const { html, ...docProps } = await renderPageWithData(Page, getAssets(), {
    initialProps,
  });
  return { html, docProps };
}

export async function iframeArticleRoute(req) {
  const lang = defined(req.params.lang, '');
  const htmlLang = getHtmlLang(lang);
  const locale = getLocaleObject(htmlLang);
  const { articleId, taxonomyId } = req.params;
  const location = { pathname: req.url };
  try {
    if (taxonomyId && taxonomyId.startsWith('urn:topic')) {
      const { html, docProps } = await doRenderPage({
        basename: lang,
        locale,
        articleId,
        isOembed: 'true',
        isTopicArticle: true,
        status: 'success',
        location,
      });

      return renderHtml(req, html, { status: OK }, docProps);
    }
    const resourceTypes = taxonomyId
      ? await fetchResourceTypesForResource(taxonomyId, htmlLang)
      : [];
    const { html, docProps } = await doRenderPage({
      resourceTypes,
      articleId,
      isOembed: 'true',
      basename: lang,
      locale,
      status: 'success',
      location,
    });

    return renderHtml(req, html, { status: OK }, docProps);
  } catch (error) {
    if (process.env.NODE_ENV !== 'unittest') {
      // skip log in unittests
      handleError(error);
    }
    const { html, docProps } = await doRenderPage({
      basename: lang,
      locale,
      location,
      status: 'error',
    });

    const status = error.status || INTERNAL_SERVER_ERROR;
    return renderHtml(req, html, { status }, docProps);
  }
}
