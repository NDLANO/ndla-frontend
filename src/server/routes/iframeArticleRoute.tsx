/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import url from 'url';
import { Request } from 'express';
import { FilledContext, HelmetProvider } from 'react-helmet-async';
import { I18nextProvider } from 'react-i18next';
import { StaticRouter } from 'react-router-dom/server.js';
import { ApolloProvider } from '@apollo/client';
import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';
import { i18nInstance } from '@ndla/ui';
import RedirectContext, { RedirectInfo } from '../../components/RedirectContext';
import config from '../../config';
import { EmotionCacheKey } from '../../constants';
import { getHtmlLang, initializeI18n, isValidLocale } from '../../i18n';
import IframePageContainer from '../../iframe/IframePageContainer';
import { InitialProps } from '../../interfaces';
import { INTERNAL_SERVER_ERROR, OK } from '../../statusCodes';
import { createApolloClient } from '../../util/apiHelpers';
import handleError from '../../util/handleError';
import { renderPageWithData, renderHtml } from '../helpers/render';

const MOCK_ASSETS = {
  'client.css': 'mock.css',
  'embed.js': 'mock.js',
  'mathJaxConfig.js': 'mock.js',
};

const assets =
  process.env.NODE_ENV !== 'unittest' && process.env.ASSETS_MANIFEST
    ? require(process.env.ASSETS_MANIFEST) //eslint-disable-line
    : MOCK_ASSETS;

if (process.env.NODE_ENV === 'unittest') {
  HelmetProvider.canUseDOM = false;
}

const getAssets = () => ({
  css: assets['client.css'],
  js: [{ src: assets['embed.js'] }],
  mathJaxConfig: { js: assets['mathJaxConfig.js'] },
});

const disableSSR = (req: Request) => {
  const urlParts = url.parse(req.url, true);
  if (urlParts.query && urlParts.query.disableSSR) {
    return urlParts.query.disableSSR === 'true';
  }
  return config.disableSSR;
};

async function doRenderPage(req: Request, initialProps: InitialProps) {
  const context: RedirectInfo = {};

  const client = createApolloClient(initialProps.locale);

  //@ts-ignore
  const helmetContext: FilledContext = {};
  const cache = createCache({ key: EmotionCacheKey });
  const noSSR = disableSSR(req);

  const i18n = initializeI18n(i18nInstance, initialProps.locale ?? config.defaultLocale);
  const Page = (
    <HelmetProvider context={helmetContext}>
      {noSSR ? (
        ''
      ) : (
        <I18nextProvider i18n={i18n}>
          <RedirectContext.Provider value={context}>
            <ApolloProvider client={client}>
              <CacheProvider value={cache}>
                <StaticRouter location={req.url}>
                  <IframePageContainer {...initialProps} />
                </StaticRouter>
              </CacheProvider>
            </ApolloProvider>
          </RedirectContext.Provider>
        </I18nextProvider>
      )}
    </HelmetProvider>
  );
  const assets = getAssets();
  const { html, ...docProps } = await renderPageWithData({
    Page,
    assets,
    disableSSR: noSSR,
    data: { initialProps },
    client,
  });
  return { html, docProps, helmetContext, redirectContext: context };
}

export async function iframeArticleRoute(req: Request) {
  const lang = req.params.lang ?? '';
  const htmlLang = getHtmlLang(lang);
  const locale = isValidLocale(htmlLang) ? htmlLang : undefined;
  const { articleId, taxonomyId } = req.params;
  try {
    const { html, docProps, helmetContext, redirectContext } = await doRenderPage(req, {
      articleId,
      taxonomyId,
      isOembed: 'true',
      basename: lang,
      locale,
      status: 'success',
    });

    return renderHtml(html, { status: redirectContext.status ?? OK }, docProps, helmetContext);
  } catch (error) {
    if (process.env.NODE_ENV !== 'unittest') {
      // skip log in unittests
      handleError(error);
    }
    const { html, docProps, helmetContext } = await doRenderPage(req, {
      basename: lang,
      locale,
      status: 'error',
    });

    const typedError = error as { status?: number };
    const status = typedError.status || INTERNAL_SERVER_ERROR;

    return renderHtml(html, { status }, docProps, helmetContext);
  }
}
