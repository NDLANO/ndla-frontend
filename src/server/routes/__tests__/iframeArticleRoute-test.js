/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { iframeArticleRoute } from '../iframeArticleRoute';

jest.mock('../../helpers/Document', () => () => (
  <html lang="en">
    <body>REPLACE_ME</body>
  </html>
));

jest.mock('../../../iframe/IframePage', () =>
  // eslint-disable-next-line react/prop-types
  ({ status }) => (
    <div>
      <p>{status}</p>
    </div>
  ),
);

test('iframeArticleRoute 200 OK', async () => {
  const response = await iframeArticleRoute({
    params: {
      lang: 'nb',
      articleId: '26050',
      taxonomyId: 'urn:resource:123',
    },
    headers: {
      'user-agent': 'Mozilla/5.0 Gecko/20100101 Firefox/58.0',
    },
    url: '/article-iframe/nb/urn:resource:123/26050',
  });

  expect(response).toMatchSnapshot();
});
