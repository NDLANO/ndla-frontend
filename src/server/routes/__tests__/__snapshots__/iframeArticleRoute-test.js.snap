// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iframeArticleRoute 200 OK 1`] = `
{
  "data": "<!doctype html><html lang="en"><body><div class="o-container"><div><p>success</p></div></div></body></html>",
  "status": 200,
}
`;
